metrics/sinks/hawkular/driver_historical.go

```diff
@@ -0,0 +1,432 @@
+}
+
+// Internal functions
+
+func (h *hawkularSink) getLabelTagValues(labelKey string, o ...metrics.Modifier) ([]string, error) {
+	tags := make(map[string]string)
+	tags[labelKey] = "*"
+	r, err := h.client.TagValues(tags, o...)
+	if r != nil {
+		return r[labelKey], err
+	}
+	return nil, err
+}
+
+func (h *hawkularSink) isNamespaceTenant() bool {
+	// This could be a bit problematic.. are we really storing the right info?
+	if h.labelTenant != "" && (h.labelTenant == core.LabelPodNamespace.Key || h.labelTenant == core.LabelPodNamespaceUID.Key) {
+		return true
+	}
+	return false
+}
+
+func (h *hawkularSink) bucketPointToAggregationValue(bp *metrics.Bucketpoint, mt *metrics.MetricType, aggregations []core.AggregationType, bucketSize time.Duration) (core.TimestampedAggregationValue, error) {
+	tav := core.TimestampedAggregationValue{
+		Timestamp:  bp.Start,
+		BucketSize: bucketSize,
+		AggregationValue: core.AggregationValue{
+			Count: &bp.Samples,
+		},
+	}
+
+	aggs := make(map[core.AggregationType]core.MetricValue, len(aggregations))
+
+	for _, a := range aggregations {
+		var val float64
+		switch a {
+		case core.AggregationTypeAverage:
+			val = bp.Avg
+		case core.AggregationTypeMaximum:
+			val = bp.Max
+		case core.AggregationTypeMinimum:
+			val = bp.Min
+		case core.AggregationTypeMedian:
+			val = bp.Median
+		case core.AggregationTypePercentile50:
+			val = bp.Median // Percentile 50 and Median are the same thing
+		case core.AggregationTypePercentile95:
+			for _, v := range bp.Percentiles {
+				if v.Quantile == 95.0 {
+					val = v.Value
+				}
+			}
+		case core.AggregationTypePercentile99:
+			for _, v := range bp.Percentiles {
+				if v.Quantile == 99.0 {
+					val = v.Value
+				}
+			}
+		}
+
+		mv := core.MetricValue{
+			MetricType: core.MetricGauge,
+			ValueType:  core.ValueFloat,
+			FloatValue: float32(val),
+		}
+		f, err := metrics.ConvertToFloat64(val)
+		if err != nil {
+			return tav, err
+		}
+		mv.FloatValue = float32(f)
+
+		aggs[a] = mv
+	}
+
+	tav.AggregationValue.Aggregations = aggs
+	return tav, nil
+}
+
+func (h *hawkularSink) datapointToMetricValue(dp *metrics.Datapoint, mt *metrics.MetricType) (core.TimestampedMetricValue, error) {
+	mv := core.MetricValue{
+		MetricType: hawkularTypeToHeapsterType(*mt),
+	}
+
+	tmv := core.TimestampedMetricValue{
+		Timestamp: dp.Timestamp,
+	}
+
+	switch *mt {
+	case metrics.Counter:
+		mv.ValueType = core.ValueInt64
+		if v, ok := dp.Value.(int64); ok {
+			mv.IntValue = int64(v)
+		}
+	case metrics.Gauge:
+		mv.ValueType = core.ValueFloat
+		f, err := metrics.ConvertToFloat64(dp.Value)
+		if err != nil {
+			return tmv, err
+		}
+		mv.FloatValue = float32(f)
+		// if v, ok := dp.Value.(float64); ok {
+
+		// }
+	}
+	tmv.MetricValue = mv
+
+	return tmv, nil
+}
+
+func (h *hawkularSink) metricNameToHawkularType(metricName string) metrics.MetricType {
+	for _, metric := range core.AllMetrics {
+		if metric.Name == metricName {
```

Comment:
This... is suboptimal.  We should probably just compose a map based on `AllMetrics`  in `core` (e.g. `core.AllMetricsByName`) or have a function, since looking up metric descriptors by name is a useful thing to be able to do and having to loop over all metrics every time is not the best.